ludwig/utils/llm_utils.py

```diff
@@ -36,3 +96,256 @@ def remove_left_padding(input_ids_sample: torch.Tensor, tokenizer: PreTrainedTok
+        tokenizer (PreTrainedTokenizer): The tokenizer used to encode the input.
+
+    Returns:
+        torch.Tensor: The attention mask tensor.
+
+    Example:
+        >>> import torch # noqa
+        >>> from transformers import PreTrainedTokenizer
+        >>> tokenizer = PreTrainedTokenizer.from_pretrained('bert-base-uncased')
+        >>> input_sentence = "This is an example sentence."
+        >>> input_ids = tokenizer.encode(input_sentence, add_special_tokens=True)
+        >>> attention_mask = create_attention_mask(input_ids, tokenizer)
+        >>> attention_mask
+        tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])
+    """
+    attention_mask = input_ids != tokenizer.pad_token_id
+    # Last token may not be padding if we've already hit the max sequence length
+    if not attention_mask[-1]:
+        # last token is padding, always attended to even if it is padding
+        attention_mask[-1] = 1
+    attention_mask = attention_mask.to(torch.int64)
+    return attention_mask
+
+
+def find_last_matching_index(tensor_a: torch.Tensor, tensor_b: torch.Tensor):
+    """Returns the last index of `tensor_a` that matches `tensor_b`. Specifically, this checks whether the tensor_b
+    is in the last tensor_b.shape[0] elements of tensor_a.
+
+    Args:
+        tensor_a (torch.Tensor): The first tensor.
+        tensor_b (torch.Tensor): The second tensor.
+
+    Returns:
+        int: The last index of `tensor_a` that matches `tensor_b`. Returns -1 if there is no matching index.
+
+    Example:
+        >>> import torch
+        >>> tensor_a = torch.tensor([1, 2, 3, 4, 5, 6, 7, 8])
+        >>> tensor_b = torch.tensor([6, 7, 8])
+        >>> last_matching_index = find_last_matching_index(tensor_a, tensor_b)
+        >>> last_matching_index
+        5
+    """
+    last_index = -1
+
+    tensor_a_length = tensor_a.shape[0]
+    tensor_b_length = tensor_b.shape[0]
+
+    # Get the last tensor_b_length elements of tensor_a.
+    tensor_a_truncated = tensor_a[-tensor_b_length:]
+
+    # Find the last matching index.
+    for i in range(tensor_b_length):
+        if torch.equal(tensor_a_truncated[i:], tensor_b[: tensor_b_length - i]):
+            last_index = tensor_a_length - tensor_b_length + i
+            break
+
+    return last_index
+
+
+def pad_target_tensor_for_fine_tuning(
+    targets: Dict[str, torch.Tensor],
+    predictions: Dict[str, torch.Tensor],
+    model_inputs: torch.Tensor,
+    of_name: str,
+) -> Dict[str, torch.Tensor]:
+    """Pad and adjust target tensors for fine-tuning LLMS models.
+
+    This function is used to pad and adjust the target tensors with -100 based on the model inputs and predictions
+    during the fine-tuning process of Language Models. Here's what this function does:
+        1. If none of the tokens from the target were in the model inputs, we create a tensor of the length of model
+            inputs with value -100s. This ignores this row from affecting loss.
```

Comment:
It's actually true for any model where we're computing CrossEntropyLoss. The main reason to use -100 is to negate/ignore those tokens from the predictions when computing loss between predictions and targets.

Specifically for LLMs, we do this to negate computing loss on predicted input tokens, and just compute loss on predicted target tokens

Reply:
Right. My question was more around why `-100` rather than `-1000` or `-101`. If it's convention, it might be good to document that somewhere, like in a constant, i.e.

```
LLM_FINE_TUNING_PAD_TOKEN_ID = -100

or 

HF_LLM_FINE_TUNING_PAD_TOKEN_ID = -100
```

Reply:
Nice! I actually named this to `IGNORE_INDEX_TOKEN_ID` all through Ludwig, hopefully that seems fair? This is definitely not specific to LLMs themselves